icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isLiteral()) {
+                result.put(option.getKey(), value.getLiteral());
+            }
+        }
+        return result;
+    }
+
+    private Map<String, Object> mf2OptToVariableOptions(Map<String, Value> options, Map<String, Object> arguments) {
+        Map<String, Object> result = new HashMap<>();
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isVariable()) {
+                result.put(option.getKey(), variableToObjectEx(value, arguments));
+            }
+        }
+        return result;
+    }
+
+    FormatterFactory getFormattingFunctionFactoryByName(Object toFormat, String functionName) {
+        // Get a function name from the type of the object to format
+        if (functionName == null || functionName.isEmpty()) {
+            if (toFormat == null) {
+                // The object to format is null, and no function provided.
+                return null;
+            }
+            Class<?> clazz = toFormat.getClass();
+            functionName = standardFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            if (functionName == null) {
+                functionName = customFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            }
+            if (functionName == null) {
+                throw new IllegalArgumentException("Object to format without a function, and unknown type: "
+                        + toFormat.getClass().getName());
+            }
+        }
+
+        FormatterFactory func = standardFunctionRegistry.getFormatter(functionName);
+        if (func == null) {
+            func = customFunctionRegistry.getFormatter(functionName);
+        }
+        return func;
+    }
+
+    String format(Map<String, Object> arguments) {
+        Pattern patternToRender;
+        List<Expression> selectors = dm.getSelectors();
+        if (selectors.isEmpty()) {
+            patternToRender = dm.getPattern();
+        } else {
+            // Collect all the selector functions in an array, to reuse
+            List<Selector> selectorFunctions = new ArrayList<>(selectors.size());
+            for (Expression selector : selectors) {
+                String functionName = selector.getFunctionName();
+                SelectorFactory funcFactory = standardFunctionRegistry.getSelector(functionName);
+                if (funcFactory == null) {
+                    funcFactory = customFunctionRegistry.getSelector(functionName);
+                }
+                if (funcFactory != null) {
+                    Map<String, Object> opt = new HashMap<>();
+                    for (Entry<String, Value> e : selector.getOptions().entrySet()) {
+                        opt.put(e.getKey(), variableToObjectEx(e.getValue(), arguments));
+                    }
```

Comment:
Question 1: This looks a bit like mf2OptToVariableOptions() except that that also checks for value.isVariable(). Should you use that helper method?

Question 2: Shouldn't we collect the fixed options at this stage? Or is this intentionally not calling the helper method so that it collects all of the options?

Reply:
Added TODO

Reply:
Done
And added unit tests.